src/servers/src/http/jaeger.rs

```
@@ -0,0 +1,1244 @@
+            }),
+        ),
+    }
+}
+
+async fn get_records(
+    sql_handler: ServerSqlQueryHandlerRef,
+    query_ctx: QueryContextRef,
+    sql: &str,
+) -> Result<Option<HttpRecordsOutput>> {
+    debug!("Executing SQL '{}' to get records", sql);
+
+    let result = handler::from_output(sql_handler.do_query(sql, query_ctx.clone()).await).await;
+
+    match result {
+        Ok(mut outputs) => {
+            if let GreptimeQueryOutput::Records(records) = outputs.0.remove(0) {
+                debug!("Get records: {:?}", records);
+                Ok(Some(records))
+            } else {
+                // If the result is not records, skip it.
+                Ok(None)
+            }
+        }
+        Err(rsp) => FailedToExecuteJaegerQuerySnafu { sql_response: rsp }.fail(),
+    }
+}
+
+// Construct Jaeger traces from records.
+// The records has column order: `trace_id`, `timestamp`, `duration_nano`, `service_name`, `span_name`, `span_id`, `span_attributes`.
+fn traces_from_records(records: HttpRecordsOutput) -> Vec<Trace> {
+    // maintain the mapping: trace_id -> (process_id -> service_name).
+    let mut trace_id_to_processes: HashMap<String, HashMap<String, String>> = HashMap::new();
+    // maintain the mapping: trace_id -> spans.
+    let mut trace_id_to_spans: HashMap<String, Vec<Span>> = HashMap::new();
+
+    // FIXME(zyy17): Should return the error if the schema is not correct.
+    for row in records.rows.into_iter() {
+        let mut span = Span::default();
+        let mut row_iter = row.into_iter();
+
+        // Set trace id.
+        if let Some(JsonValue::String(trace_id)) = row_iter.next() {
+            span.trace_id = trace_id.clone();
+            trace_id_to_processes.entry(trace_id).or_default();
+        }
+
+        // Convert timestamp from nanoseconds to microseconds.
+        if let Some(JsonValue::Number(timestamp)) = row_iter.next() {
+            span.start_time = timestamp.as_u64().unwrap_or(0) / 1000;
```

Do you think if we need to raise an error here instead of using default 0?

---

Sounds reasonable.